plugins/postgres/dbt/include/postgres/macros/utils/columns_spec_ddl.sql

```diff
@@ -0,0 +1,23 @@
+{% macro postgres__get_columns_spec_ddl() %}
+  {# loop through user_provided_columns to create DDL with data types and constraints #}
+    {%- set user_provided_columns = model['columns'] -%}
+    (
+    {% for i in user_provided_columns %}
+      {% set col = user_provided_columns[i] %}
+      {% set constraints = col['constraints'] %}
+      {% set constraints_check = col['constraints_check'] %}
+      {{ col['name'] }} {{ col['data_type'] }} {% for x in constraints %} {{ x or "" }} {% endfor %} {% if constraints_check -%} check {{ constraints_check or "" }} {%- endif %} {{ "," if not loop.last }}
```

Comment:
I vote option 1 because it has less database performance hits because option 2 using subqueries depends a lot more on the database to optimize it(which varies). 

I'll make this change!

Reply:
I'm not sure we create any temp tables in the current implementation, so where would we do the mid-transaction check in option 1?

Reply:
@dave-connors-3 The way we replace preexisting views/tables on Postgres + Redshift is by creating a "temp" relation (with a suffix `__dbt_tmp`, not always _actually_ a temporary table), and then `alter rename` swapping the existing/old + new relations, before dropping the old one. That ensures zero downtime, and prevents the view/table from being missing if the model SQL runs into an error.

So we _could_ try to stick another check in the middle of that process. But I actually think `get_columns_in_query` might be preferable across the board, because it will also catch mismatches _before_ we run a potentially expensive query to create the table with new data.

Reply:
I made the change by raising a compile error if the names and order of the columns don't match exactly in a global macro. I added 2 tests to prove this. 

![image](https://user-images.githubusercontent.com/19176976/212407382-4633226b-df52-4ca2-892b-68e28710c72c.png)